### PR TITLE
Reset bgfx and camera projection on window resize

### LINKED ISSUE #100: Reset projection matrix on window resize

### PR DESCRIPTION
**Pull request by captnfab**
On window resize events, perform a [bgfx::reset](https://bkaradzic.github.io/bgfx/bgfx.html) to reset framebuffers, and reset camera projection.

Fixes #100, #211

### CODE REVIEW

File: src/Game.cpp [hunk 1]
```diff
@@ -179,6 +179,15 @@ bool Game::ProcessEvents(const SDL_Event& event)
 		{
 			return false;
 		}
+		else if (event.window.event == SDL_WINDOWEVENT_SIZE_CHANGED)
+		{
+			int width, height;
```
Comment (bwrsandman) [inline src/Game.cpp:184]: I know we do it in other places but for readability, don't declare two variables on the same line.
If you really want one line try this:
```
glm::ivec2 resolution;
_window->GetSize(resolution,x, resolution.y);
```

File: src/Game.cpp [hunk 2]
```diff
@@ -179,6 +179,15 @@ bool Game::ProcessEvents(const SDL_Event& event)
...
+			_window->GetSize(width, height);
+			auto aspect = _window->GetAspectRatio();
+			
+			bgfx::reset(width, height, BGFX_RESET_NONE);
+			_camera->SetProjectionMatrixPerspective(70.0f, aspect, 1.0f, 65536.0f);
```
Comment (bwrsandman) [inline src/Game.cpp:189]: These magic numbers should be set as named constants and eventually should come from game data or config file.
I would recommend you make an entry in the Config struct and use it in all the places we set FOV and near and far.

**Review by bwrsandman — CHANGES_REQUESTED** [review]

**Review by bwrsandman — APPROVED** [review]

File: src/Game.cpp (hunk 2 above)
Comment (captnfab) [reply src/Game.cpp]: I added three new parameters in Config struct. These are used at least at two point, at camera initialization, and on resize.
I used your static_cast presentation.

**Comment by captnfab** [comment]
@bwrsandman : thanks for your reviews!
Is there some reference document for this project's coding style?

**Comment by bwrsandman** [comment]
There isn't one yet. In general, I would say try to match the style of the files you're editing.
We have a clang-format formatting checker and config file.
We check for copyright also.
There's also a clang-tidy config but we don't yet have automatic checking for that #221.

One thing I often check for is that you follow the [google style order of header includes](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes) and that headers do declarations and not implementations.

Special care for ECS code: Use Archtetypes instead of polymorphism (classes), Entities are plain old data and Systems do the logic.

**Review by bwrsandman — CHANGES_REQUESTED** [review]
I tested it and it doesn't seem to work very well

Scaling up
Screenshot from 2022-04-19 15-43-14

Scaling Down
Screenshot from 2022-04-19 15-43-22


After a few resizes I get this followed by a crash:
Screenshot from 2022-04-19 15-43-35

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  bgfx: vcpkg/buildtrees/bgfx/src/5889d79ede-d8473b23d1.clean/src/renderer_vk.cpp:7830: FATAL (DebugCheck): vkQueueSubmit(m_queue, 1, &si, m_completedFence); VK error 0xfffffffc: VK_ERROR_DEVICE_LOST
[1]    482260 IOT instruction (core dumped)  ./cmake-build-presets/ninja-multi-vcpkg/bin/Debug/openblack -g 
```

**Comment by captnfab** [comment]
Mmmm, there seems to be something weird happening when width is larger than 1280px.

**Comment by captnfab** [comment]
PR still not ready for merge.

**Comment by captnfab** [comment]
@bwrsandman : I haven't been able to reproduce your crash, but I fixed a couple of bugs that could have been responsible for it.

**Comment by captnfab** [comment]
There's still one problem. To do the reset, I need to know the flags to apply.
I need to know if vsync needs to be used or not.

I'd say the vsync flag has to be saved somewhere in the Redrerer class, and the bgfx::reset needs to be moved to a Renderer::reset method.

**Comment by raffclar** [comment]
> There's still one problem. To do the reset, I need to know the flags to apply. I need to know if vsync needs to be used or not.
> 
> I'd say the vsync flag has to be saved somewhere in the Redrerer class, and the bgfx::reset needs to be moved to a Renderer::reset method.

You can see the reset we use here:
https://github.com/openblack/openblack/blob/master/src/Renderer.cpp#L142
```c++
	init.resolution.reset = BGFX_RESET_NONE;
	if (vsync)
	{
		init.resolution.reset |= BGFX_RESET_VSYNC;
	}
```

Looks like we'll need to remember if vsync has been turned on or off in `Renderer`.

**Comment by bwrsandman** [comment]
New changes have fixed it.

**Comment by bwrsandman** [comment]
vsync comes from `args.vsync` in `Game`

File: src/Gui/Gui.cpp [hunk 3]
```diff
@@ -209,6 +209,12 @@ bool Gui::ProcessEventSdl2(const SDL_Event& event)
 		io.KeySuper = ((SDL_GetModState() & KMOD_GUI) != 0);
 		return io.WantCaptureKeyboard;
 	}
+	case SDL_WINDOWEVENT:
+		if (event.window.event == SDL_WINDOWEVENT_SIZE_CHANGED)
+		{
+			bgfx::reset(static_cast<uint32_t>(event.window.data1), static_cast<uint32_t>(event.window.data2), BGFX_RESET_VSYNC);
```
Comment (bwrsandman) [inline src/Gui/Gui.cpp:215]: Why does this have to be caught in the Gui? This is a bad place to have this code.

Comment (bwrsandman) [reply src/Gui/Gui.cpp:215]: I've tested resizing without this and it doesn't contribute.

Comment (captnfab) [reply src/Gui/Gui.cpp:215]: In the event loop (https://github.com/openblack/openblack/blob/ddf190ae27a2779adb260c2719b87d09f8fbb0ce/src/Game.cpp#L131) the Gui is the first to handle SDL events.

By default, if an event is out of the window it is dropping it (https://github.com/openblack/openblack/blob/ddf190ae27a2779adb260c2719b87d09f8fbb0ce/src/Gui/Gui.cpp#L215)
The problem in this case is that the window cannot be scaled up.

So an important part in the event handling in Gui.cpp is the `return false`.

However, as you said, the bgfx::reset is useless. It's already present in the Game::ProcessEvents function. I'm removing it from the Gui.

Comment (captnfab) [reply src/Gui/Gui.cpp]: All that remains to do is store the reset flags in the renderer and reuse them during reset.

File: src/Gui/Gui.cpp [hunk 4]
```diff
@@ -209,6 +209,11 @@ bool Gui::ProcessEventSdl2(const SDL_Event& event)
 		io.KeySuper = ((SDL_GetModState() & KMOD_GUI) != 0);
 		return io.WantCaptureKeyboard;
 	}
+	case SDL_WINDOWEVENT:
+		if (event.window.event == SDL_WINDOWEVENT_SIZE_CHANGED)
+		{
+			return false;
+		}
 	}
```
Comment (bwrsandman) [inline src/Gui/Gui.cpp:217]: Missing a `break `

Comment (captnfab) [reply src/Gui/Gui.cpp:218]: Fixed.

File: src/Gui/Gui.cpp (hunk 3 above)
Comment (captnfab) [reply src/Gui/Gui.cpp]: I added a uint32_t _bgfxReset member in Renderer for saving the reset flags. This should solve the last issue we had.

**Comment by bwrsandman** [comment]
LGTM. Just want to test it when I get home

**Comment by bwrsandman** [comment]
It starts up like this without even resizing
Screenshot from 2022-04-20 20-53-37

**Comment by captnfab** [comment]
> It starts up like this without even resizing. After resizing it's okay.
I can't reproduce this bug on my setups, but it must be related to the addition of `bgfx::reset(width, height, _bgfxReset);` in the `Renderer::ConfigureView` method.
I'll add a Renderer::Reset method and move the bgfx::reset code there, so that it's not called on init but only after resize.

**Comment by raffclar** [comment]
> It starts up like this without even resizing. After resizing it's okay. Screenshot from 2022-04-20 20-53-37

Same issue but other than that works fine on Windows

**Comment by captnfab** [comment]
@bwrsandman @raffclar : I believe the bug is now fixed.

I wasn't able to reproduce it, it might be related to the version of bgfx we are using.(I am using rev cae7c83f896d2b4740ce232f6224da16b949b224)

**Review by bwrsandman — APPROVED** [review]
Tested it and it works great

**Pull request merged**